presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java

```diff
@@ -1111,6 +1113,15 @@ public void validateScan(Session session, TableHandle table)
                             projections.size(),
                             table);
 
+                    Set<String> assignedVariables = newAssignments.stream()
+                            .map(ProjectionApplicationResult.Assignment::getVariable)
+                            .collect(toImmutableSet());
+                    newProjections.stream()
+                            .flatMap(connectorExpression -> ConnectorExpressions.extractVariables(connectorExpression).stream())
+                            .map(Variable::getName)
+                            .filter(variableName -> !assignedVariables.contains(variableName))
+                            .findAny().ifPresent(variableName -> { throw new IllegalStateException("Unbound variable: " + variableName); });
```

Comment:
Throw a proper PrestoException

Reply:
i thought about that.

this should happen only during connector development, so i decided to go "lite" about these exceptions, and use brief forms of verification (like `verify`)
(eg if we were to convert all `requireNonNull` to `if ... throw PrestoException` would not be fun, and would greatly harm readability for little or no value to end users)